# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1744494034 -7200
# Parent  bef74dcc0832b7a288b0b3b1da5af5cf553138af
Feature: Set startup focus based on pref. mail.startup.focus (bug 425025, bug 1620310).

diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -1463,8 +1463,15 @@ pref("mail.multiselect_message_shows_cur
 pref("mail.startupMinimized", false);
 // Avoid .part files on desktop when opening attachments from compose window.
 pref("browser.download.start_downloads_in_tmp_dir", true);
 
+// Set the focus to a defined element on startup:
+// 0: no focus
+// 1: spaces toolbar (collapsed: button, expanded: mail icon)
+// 2: folder pane (if any, falls back to 1)
+// 3: thread pane (if any, falls back to 1).
+pref("mail.startup.focus", 0);
+
 // Unified toolbar
 
 // 0: icons beside text
 // 1: icons above text
diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -210,8 +210,44 @@ window.addEventListener("DOMContentLoade
   CalMetronome.on("day", refreshGroupedBySortView);
 
   updateZoomCommands();
 
+  // Now that we restored everything, let's set the focus.
+  const whereToFocus = Services.prefs.getIntPref("mail.startup.focus", 0);
+  if (whereToFocus <= 0 || whereToFocus > 3) {
+    return;
+  }
+
+  if (
+    top.window.document.getElementById("tabmail")?.currentTabInfo.mode.name ==
+    "mail3PaneTab"
+  ) {
+    if (whereToFocus == 2) {
+      folderTree.focus();
+      return;
+    } else if (whereToFocus == 3) {
+      const table = threadTree.querySelector("[is='tree-view-table-body']");
+      // If "Account Central" is shown, the width is 0.
+      if (table.clientWidth > 0) {
+        table.focus();
+      } else {
+        folderTree.focus();
+      }
+      return;
+    }
+  }
+
+  // OK, let's try focusing the spaces toolbar.
+  let button = top.window.document.getElementById("spacesPinnedButton");
+  if (button && !button.hidden) {
+    button.focus();
+    return;
+  }
+  button = top.window.document.getElementById("mailButton");
+  if (button && !button.hidden) {
+    button.focus();
+  }
+
   // Update the state of the about:3pane being fully loaded.
   hasDOMContentLoaded.resolve();
 });
 
